core/src/Cabana_VerletList.hpp

```diff
@@ -375,6 +382,108 @@ struct VerletListBuilder
             } );
     }
 
+    // Neighbor count team (+serial) operator (only used for CSR lists).
+    KOKKOS_INLINE_FUNCTION
+    void
+    operator()( const CountNeighborsTag<TeamOpTag> &,
+                const typename CountNeighborsPolicy::member_type &team ) const
+    {
+        // The league rank of the team is the cardinal cell index we are
+        // working on.
+        int cell = team.league_rank();
+
+        // Get the stencil for this cell.
+        int imin, imax, jmin, jmax, kmin, kmax;
+        cell_stencil.getCells( cell, imin, imax, jmin, jmax, kmin, kmax );
+
+        // Operate on the particles in the bin.
+        std::size_t b_offset = bin_data_1d.binOffset( cell );
+        Kokkos::parallel_for(
+            Kokkos::TeamThreadRange( team, 0, bin_data_1d.binSize( cell ) ),
+            [&]( const int bi ) {
+                // Get the true particle id. The binned particle index is the
+                // league rank of the team.
+                std::size_t pid = linked_cell_list.permutation( bi + b_offset );
+
+                if ( ( pid >= pid_begin ) && ( pid < pid_end ) )
+                {
+                    // Cache the particle coordinates.
+                    double x_p = position( pid, 0 );
+                    double y_p = position( pid, 1 );
+                    double z_p = position( pid, 2 );
+
+                    // Loop over the cell stencil.
+                    int stencil_count = 0;
+                    for ( int i = imin; i < imax; ++i )
+                        for ( int j = jmin; j < jmax; ++j )
+                            for ( int k = kmin; k < kmax; ++k )
+                            {
+                                // See if we should actually check this box for
+                                // neighbors.
+                                if ( cell_stencil.grid.minDistanceToPoint(
+                                         x_p, y_p, z_p, i, j, k ) <= rsqr )
+                                {
+                                    // Check the particles in this bin to see if
+                                    // they are neighbors. If they are add to
+                                    // the count for this bin.
+                                    int cell_count = 0;
+                                    std::size_t a_offset =
+                                        linked_cell_list.binOffset( i, j, k );
+
+                                    for ( int n = 0;
+                                          n <
+                                          linked_cell_list.binSize( i, j, k );
+                                          n++ )
+                                    {
+                                        //  Get the true id of the candidate
+                                        //  neighbor.
+                                        std::size_t nid =
```

Comment:
It would be nice to abstract this part so we don't have to repeat all of this code. The loop body is mostly the same - just different how the loop is called (serial vs. vector). Can we separate the loop body into another function and abstract how we call the loop? That should let us combine a lot of code. This function is then templated on the type tag and the loop function will have explicit instantiations.

Reply:
Agreed - fixed this